### ISSUE TITLE
Mutual nearest neighbors

### ISSUE THREAD

**Issue by ChrisLLLLL** [zju3dv/LoFTR#14]
Hello,in coarse_matching.py ,How MNN is implemented？I think it’s a bit difficult for me to understand just by looking at the code Where did this method have been proposed before, and I didn’t find a related introduction. Or is it your new method.
thank you！

**Comment by zehongs** [comment]
```python
# 2. mutual nearest
mask = mask \
    * (conf_matrix == conf_matrix.max(dim=2, keepdim=True)[0]) \
    * (conf_matrix == conf_matrix.max(dim=1, keepdim=True)[0])
```
Find the max along each column or row. And the code above gives you the MNN-mask.

```
# 3. find all valid coarse matches
# this only works when at most one `True` in each row
mask_v, all_j_ids = mask.max(dim=2)
b_ids, i_ids = torch.where(mask_v)
j_ids = all_j_ids[b_ids, i_ids]
mconf = conf_matrix[b_ids, i_ids, j_ids]
```
This is an equivalent way of `torch.where(mask)` if and only if at most one `True` in the last dim. The equivalent way reduces GPU memory to a large extent. I haven't checked the implementation of `torch.where()`, but from my experience, I guess it first creates a N*L*S*3*long tensor as full-indices, and then uses mask to select indices. In our case, e.g. 640*480 pair + 1/8 resolution, 1*4800*4800*3*(4/1e6)=276MB v.s. 1*4800*2*(4/1e6)=0.04MB.
